utils.go

```diff
@@ -122,52 +124,78 @@ const defaultSymFile = "/proc/kallsyms"
 // for older syscall functions to run on newer kernels
 func getSyscallName(name string, symFile string) (string, error) {
 	// Get kernel symbols
-	syms, err := os.ReadFile(symFile)
+	syms, err := os.Open(symFile)
 	if err != nil {
 		return "", err
 	}
-	return getSyscallFnNameWithKallsyms(name, string(syms))
+	defer syms.Close()
+
+	return getSyscallFnNameWithKallsyms(name, syms, "")
 }
 
-func getSyscallFnNameWithKallsyms(name string, kallsymsContent string) (string, error) {
-	var arch string
-	switch runtime.GOARCH {
-	case "386":
-		arch = "ia32"
-	case "arm64":
-		arch = "arm64"
-	default:
-		arch = "x64"
+func getSyscallFnNameWithKallsyms(name string, kallsymsContent io.Reader, arch string) (string, error) {
+	if arch == "" {
+		switch runtime.GOARCH {
+		case "386":
+			arch = "ia32"
+		case "arm64":
+			arch = "arm64"
+		default:
+			arch = "x64"
+		}
 	}
 
 	// We should search for new syscall function like "__x64__sys_open"
 	// Note the start of word boundary. Should return exactly one string
-	regexStr := `(\b__` + arch + `_[Ss]y[sS]_` + name + `\b)`
-	fnRegex := regexp.MustCompile(regexStr)
+	newSyscall := regexp.MustCompile(`\b__` + arch + `_[Ss]y[sS]_` + name + `\b`)
+	// If nothing found, search for old syscall function to be sure
+	oldSyscall := regexp.MustCompile(`\b[Ss]y[sS]_` + name + `\b`)
+	// check for '__' prefixed functions, like '__sys_open'
+	prefixed := regexp.MustCompile(`\b__[Ss]y[sS]_` + name + `\b`)
 
-	match := fnRegex.FindAllString(kallsymsContent, -1)
-	if len(match) > 0 {
-		return match[0], nil
+	patterns := []struct {
+		pattern *regexp.Regexp
+		result  string
+	}{
+		{newSyscall, ""},
+		{oldSyscall, ""},
+		{prefixed, ""},
 	}
 
-	// If nothing found, search for old syscall function to be sure
-	regexStr = `(\b[Ss]y[sS]_` + name + `\b)`
-	fnRegex = regexp.MustCompile(regexStr)
-	match = fnRegex.FindAllString(kallsymsContent, -1)
-	// If we get something like 'sys_open' or 'SyS_open', return
-	// either (they have same addr) else, just return original string
-	if len(match) > 0 {
-		return match[0], nil
+	scanner := bufio.NewScanner(kallsymsContent)
+	scanner.Split(bufio.ScanLines)
+
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		if !strings.Contains(line, name) {
+			continue
+		}
+
+		for i, p := range patterns {
+			// if we already found a match for this pattern we continue
+			if p.result != "" {
+				continue
```

Comment:
One way to approach this would be to keep a slice of the patterns to search, and remove it once you found a result. Then you can break out once `len(patternsToSearch) == 0`. You'd have to keep the regexp->result mapping elsewhere though.

Reply:
Sounds good.